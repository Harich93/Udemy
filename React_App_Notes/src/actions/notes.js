
import Swal from "sweetalert2";
import { db } from "../firebase/firebaseConfig";
import { fileUpload } from "../helpers/fileUpload";
import { loadNotes } from "../helpers/loadNotes";
import { types } from "../types/types";

export const stratNewNote = () => {
    return async ( dispatch, getState ) => {

        const { uid } = getState().auth;

        const newNote = {
            title: '',
            body: '',
            date: new Date().getTime()
        }

        const doc = await db.collection( `${ uid }/journal/notes` ).add( newNote )
        
        dispatch( activeNote( doc.id, newNote) );
        dispatch( addNewNote( doc.id, newNote ) );
    }

}
 
export const activeNote = ( id, note ) => ({
    type: types.notesActive,
    payload: {
        id,
        ...note
    }
});

export const addNewNote =( id, note ) => ({
    type: types.notesAddNew,
    payload: {
        id,
        ...note
    }
})

export const startLoadingNotes = ( uid ) => {
    return async( dispatch ) => {
        const notes = await loadNotes( uid );
        dispatch( setNotes( notes ) );
    }
}

export const setNotes = ( notes ) => ({
    type: types.notesLoad,
    payload: notes
});

export const savedNote = ( note ) => {

    return async( dispatch, getState ) => {

        const { uid } = getState().auth;
        const noteToFirestore = { ...note };
        delete noteToFirestore.id;
        // !note.url && delete note.url
        

        const arch = await db.doc( `${ uid }/journal/notes/${ note.id }` )
            .update( noteToFirestore )

        console.log( arch )

        dispatch( refreshNote( note.id, note ) );
        Swal.fire( 'Saved', note.title, 'success');
    } 
}

export const refreshNote = ( id, note ) => ({
    type: types.notesUpdated,
    payload: {
        id,
        note: {
            id,
            ...note
        }
    }
})

export const startUploading = ( file ) => {
    return async( dispatch, getState ) => {
        const { active: activeNote } = getState().note;
        Swal.fire({
            title: 'Uploading...',
            text: 'Please await...',
            allowOutsideClick: false,
            showConfirmButton: false,
            willOpen: () => {
                Swal.showLoading();
            },   
        });
        
        const fileUrl = await fileUpload( file );
        activeNote.url = fileUrl;
        
        dispatch( savedNote( activeNote ) );

        Swal.close();
    }
}

export const startDeleting = ( id ) => {
    console.log(id)
    return async( dispatch, getState ) => {
        
        const { uid } = getState().auth;
        
        await db.doc( `${ uid }/journal/notes/${ id }` ).delete()
            .then( () => Swal.fire( 'Delete', id, 'success') )
            .catch( console.log )

        dispatch( deleteNote(id) ); 
    }

}

export const deleteNote = ( id ) => ({
    type: types.notesDeleted,
    payload: id
})

export const logoutNote = () => ({
    type: types.notesLogoutCleaning
})