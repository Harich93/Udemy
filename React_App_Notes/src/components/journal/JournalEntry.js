import moment from 'moment'
import React from 'react'
import { useDispatch } from 'react-redux';
import { activeNote } from '../../actions/notes';


export const JournalEntry = ( note ) => {

    const {id, body, date, title, url } = note

    const dispatch = useDispatch()

    const noteDate = moment(date);
    
    const handleEditNote = () => {
        dispatch( activeNote( id, note ) );
    }

    return (
        <div 
            className='journal__entry pointer animate__animated animate__fadeIn'
            onClick={ handleEditNote }
        >

            {
                url &&
                    <div
                        className='journal__entry-picture'
                        style={ { 
                            backgroundImage: `URL(${ url })`, 
                            backgroundSize: 'cover'
                        } }
                    ></div>
            }


                <div className='journal__entry-body'>
                    <p className='journal__entry-title'>  
                        { title }
                    </p>
                    <p className='journal__entry-content'>  
                        { body }
                    </p>
                </div>

                <div className='journal__entry-date-box'>
                    <span>{ noteDate.format('dddd') }</span>
                    <h4>{ noteDate.format('Do') }</h4>
                </div>

        </div>
    )
}
