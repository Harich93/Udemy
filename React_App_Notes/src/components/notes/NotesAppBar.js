import moment from 'moment'
import React from 'react'
import { useDispatch, useSelector } from 'react-redux';
import { savedNote, startUploading } from '../../actions/notes';

export const NotesAppBar = () => {
    
    const date = moment().format( 'LL' );

    const note = useSelector(state => state.note.active );
    
    const dispatch = useDispatch();
    
    const handleSave = () => {

        dispatch( savedNote( note ) );
    }

    const handleUpdatePicture = () => {
        document.querySelector('#fileSelector').click();
    }
    
    const handleFileChanged = ( e ) => {
        const file = e.target.files[0];

        file && dispatch( startUploading( file ) );

    }

    return (
        <div className='notes__appbar'>
            <span>{ date }</span>

            <input
                id='fileSelector'
                name='file'
                type='file'
                style={ { display: 'none' } }
                onChange={ handleFileChanged }
            />

            <div>
                <button 
                    className='btn'
                    onClick={ handleUpdatePicture }    
                > 
                    Picture
                </button>

                <button 
                    className='btn'
                    onClick={ handleSave }
                > 
                    Save
                </button>
            </div>
        
        </div>
    )
}
