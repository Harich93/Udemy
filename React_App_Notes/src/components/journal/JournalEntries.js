import React from 'react'
import { useSelector } from 'react-redux'
import { JournalEntry } from './JournalEntry'

export const JournalEntries = () => {
    
    const { notes }  = useSelector(state => state.note)

    return (
        <div className='journal__entries  animate__animated animate__fadeInLeft animate__slow'>
            {
                notes.map( note => (
                    <JournalEntry
                        key={ note.id } 
                        { ...note }
                    />
                ))
            }
        </div>
    )
}
