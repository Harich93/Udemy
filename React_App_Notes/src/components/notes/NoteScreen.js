import React, { useEffect, useRef } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { activeNote, startDeleting } from '../../actions/notes';
import { useForm } from '../../hooks/useForm';
import { NotesAppBar } from './NotesAppBar'

export const NoteScreen = () => {

    const dispatch = useDispatch();

    const note = useSelector(state => state.note.active );
    const [ values, handleInputChange, reset ] = useForm(note);
    const { title, body, url } = values;

    const activeId = useRef( note.id );
    const activeUrl = useRef( note.url );

    const handleDelete = () => {
        dispatch( startDeleting( note.id ) );
    }
    
    useEffect(() => {

        if ( note.id !== activeId.current ) {
            reset( note );
            activeId.current = note.id;
        }
        if ( note.url !== activeUrl.current ) {
            reset( note );
            activeUrl.current = note.url;
        }

    }, [note, reset, note.url ]);

    useEffect(() => {
        
        dispatch( activeNote( values.id, { ...values } ) );

    }, [values, dispatch,] )
    
    return (
        <div className='notes___main-content'>
            <NotesAppBar />

            <div className='notes__content'>

                    <input 
                        type='text'
                        name='title'
                        placeholder='Some awasome title'
                        className='notes__title-input'
                        value={ title }
                        autoComplete='off'
                        onChange={ handleInputChange }
                    />

                    <textarea
                        placeholder='What happened today'
                        name='body'
                        className='notes__textarea'
                        value={ body }
                        onChange={ handleInputChange }
                    ></textarea>

                    {
                        url &&
                            <div className='notes__image'>
                                <img
                                    src={ url }
                                    alt='Imagen nota'
                                />
                            </div>
                    }

             
            </div>
            <button
                className='btn btn-delete'
                onClick={ handleDelete }
            >Delete</button>
        </div>
    )
}
