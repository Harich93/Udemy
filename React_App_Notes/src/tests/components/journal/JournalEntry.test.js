import '../../../setupTest'

import React from 'react';
import thunk from 'redux-thunk';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { mount } from 'enzyme';

import { JournalEntry } from '../../../components/journal/JournalEntry';
import { activeNote } from '../../../actions/notes';

jest.mock('../../../actions/notes', () => ({
    activeNote: jest.fn()
})) 

const middlewares = [thunk];
const mockStore = configureStore( middlewares );

const initialState = {
    auth:{},
    note:{},
    ui:{},
}

const nota = {
    id: '123',
    body: 'Body',
    title: 'Title',
    url: 'http://urlImagen.jpg'
}

const store = mockStore( initialState );
store.dispatch = jest.fn();

const wrapper = mount(
    <Provider store={ store } >
        <JournalEntry { ...nota } />
    </Provider>
)

describe('Pruebas en <JournalEntry />', () => {
    
    test('Deme mostrarse correctamente', () => {
        expect( wrapper ).toMatchSnapshot();
    });

    test('Debe de llamar activeNote', () => {
        wrapper.find('.journal__entry').prop('onClick')();

        expect( activeNote ).toHaveBeenCalledWith('123', nota)
    });

});