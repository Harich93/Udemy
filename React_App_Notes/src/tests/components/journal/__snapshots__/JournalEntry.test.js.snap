// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <JournalEntry /> Deme mostrarse correctamente 1`] = `
<div
  className="journal__entry pointer animate__animated animate__fadeIn"
  onClick={[Function]}
>
  <div
    className="journal__entry-picture"
    style={
      Object {
        "backgroundImage": "URL(http://urlImagen.jpg)",
        "backgroundSize": "cover",
      }
    }
  />
  <div
    className="journal__entry-body"
  >
    <p
      className="journal__entry-title"
    >
      Title
    </p>
    <p
      className="journal__entry-content"
    >
      Body
    </p>
  </div>
  <div
    className="journal__entry-date-box"
  >
    <span>
      Wednesday
    </span>
    <h4>
      4th
    </h4>
  </div>
</div>
`;
